app/src/main/java/com/kidssaveocean/fatechanger/firebase/FirebaseService.kt

```diff
@@ -0,0 +1,80 @@
+package com.kidssaveocean.fatechanger.firebase
+
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.FirebaseDatabase
+import com.google.firebase.database.ValueEventListener
+import java.util.*
+
+class FirebaseService : Observable ()  {
+
+    private var auth: FirebaseAuth = FirebaseAuth.getInstance()
+    private val countries : MutableCollection<CountryModel> = mutableListOf()
+
+    var hasCountries = false
+        get() = countries.isNotEmpty()
+
+    companion object {
+
+        private val instance: FirebaseService = FirebaseService()
+        /* The 'table' in Firebase database */
+        @JvmField val COUNTRIES_TABLE = "COUNTRIES"
+        /* Needed properties in that table */
+        @JvmField val COUNTRY_NAME = "country_name"
+        @JvmField val COUNTRY_ADDRESS = "country_address"
+        @JvmField val COUNTRY_HEAD_OF_STATE_TITLE = "country_head_of_state_title"
+        @JvmField val LATITUDE = "latitude"
+        @JvmField val LONGITUDE = "longitude"
+        @JvmField val LETTERS_WRITTEN_TO_COUNTRY = "letters_written_to_country"
+
+        @Synchronized
+        fun getInstance(): FirebaseService {
+            return instance
+        }
+    }
+
+    init {
+        auth.signInAnonymously().addOnCompleteListener { task ->
+            if (task.isSuccessful) {
+                var dbInstance = FirebaseDatabase.getInstance().reference;
+                dbInstance.child(COUNTRIES_TABLE).addListenerForSingleValueEvent(object : ValueEventListener {
+
+                    override fun onDataChange(dataSnapshot: DataSnapshot) {
+                        for (item in dataSnapshot.children) {
+                            val countryModel = CountryModel.create()
+                            for (property in item.children) {
+                                try {
+                                    when (property.key) {
+                                        COUNTRY_NAME -> countryModel.countryName = property?.value as String
```

Comment:
You might be able to use kotlin json parsing here to avoid lots of manual parsing.

Reply:
Can you clarify something? Firebase's package in the project returns me a nested HashMaps. I don't see a valid json object as the response. Should I convert it into json and deserialize it back as collection of country models? Am I missing something? It just does not feel right. toString() of the response is not a valid json also.